### PR TITLE
Search

### PR DESCRIPTION
**Pull request by BohdanBusko**
**STATUS**
READY

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [ ] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs
screenshot from 2018-11-12 19-54-06
screenshot from 2018-11-12 19-53-51

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -1,18 +1,7 @@
...
-// about supported directives.
-//
 //= require jquery3
 //= require jquery_ujs
 //= require activestorage
 //= require turbolinks
-//= require_tree .
-//= require jquery
+//= require search
+//= require dashboard
```
Comment (romaslmd) [inline app/assets/javascripts/application.js:7]: isnt this file removed?

File: app/assets/javascripts/search.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
+document.addEventListener('DOMContentLoaded', e => {
+  const workspace_url = 'http://localhost:3000/account/workspaces.json';
```
Comment (romaslmd) [inline app/assets/javascripts/search.js]: coffe? ES5+ isnt really supported by some browsers

File: app/assets/javascripts/search.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    .then(responce => {
+      if (responce[objectName].length > 0) {
+        displayObject.style.display = 'block';
+        innerObject.innerHTML = jsonToHTML(responce[objectName], title);
+      } else {
+        innerObject.innerHTML = '';
+        displayObject.style.display = 'none';
+      }
+    })
+    .catch(error => { console.log(error) });
```
Comment (romaslmd) [inline app/assets/javascripts/search.js]: end user flash?

**Comment by romaslmd** [comment]
tests?

File: app/assets/javascripts/search.js (hunk 2 above)
Comment (romaslmd) [reply app/assets/javascripts/search.js]: remove base url

File: app/assets/javascripts/search.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
+document.addEventListener('DOMContentLoaded', e => {
+  const workspace_url = 'http://localhost:3000/account/workspaces/search.json';
+  const projects_url = 'http://localhost:3000/account/projects/search.json';
+  const tasks_url = 'http://localhost:3000/account/tasks/search.json';
```
Comment (romaslmd) [inline app/assets/javascripts/search.js]: pls, use dynamic relative paths from rails url helpers

File: app/assets/javascripts/search.js (hunk 3 above)
Comment (BohdanBusko) [reply app/assets/javascripts/search.js]: what do you mean?

File: app/assets/javascripts/search.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    var workspacesBlock = document.getElementById('workspaces-block');
+    var projectsBlock = document.getElementById('projects-block');
+    var tasksBlock = document.getElementById('tasks-block');
+
+    var timer;
+
+    clearTimeout(timer)
+    timer = setTimeout(() => {
+      if (e.target.value) {
+        getData(Routes.account_workspaces_search_path({ format: "json" }),
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/search.js]: why 3 calls?

Only one call is needed, which is doing Search on all models

File: app/assets/javascripts/search.js (hunk 3 above)
Comment (igorkasyanchuk) [reply app/assets/javascripts/search.js]: he means that if any error how a user will be notified?

File: app/models/project.rb [hunk 6]
```diff
@@ -28,6 +28,7 @@ class Project < ApplicationRecord
   has_many :users, through: :user_projects
 
   scope :order_desc, -> { order(id: :desc) }
+  scope :search_projects, -> (search) { where("name ~* '\\m#{search}'").order(name: :asc) }
```
Comment (AndriySolonyna) [inline app/models/project.rb]: this code vulnerable to sql injections

File: app/assets/javascripts/search.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+          workspaces,
+          'workspaces',
+          'WORKSPACES',
+          e.target.value,
+          workspacesBlock
+        );
+
+        getData(Routes.account_projects_search_path({ format: "json" }),
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/search.js]: remove

only 1 API call

implement SearchController and put logic there

File: app/models/project.rb (hunk 6 above)
Comment (igorkasyanchuk) [inline app/models/project.rb]: sql injections

File: app/models/user.rb [hunk 8]
```diff
@@ -87,6 +87,18 @@ def watching?(task)
     self.tasks.where(id: task.id).exists?
   end
 
+  def search_tasks(search)
+    tasks = Array.new
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: remove this array `tasks` and user `collect` or `map`

File: config/routes.rb [hunk 9]
```diff
@@ -7,6 +7,9 @@
 
   namespace :account do
     get '/dashboard', to: 'dashboard#index'
+    get '/workspaces/search', to: 'workspaces#search', defaults: { format: :json }
```
Comment (igorkasyanchuk) [inline config/routes.rb]: one route only

File: spec/controllers/account/projects_controller_spec.rb [hunk 10]
```diff
@@ -69,6 +69,13 @@
       end
     end
 
+    describe 'GET /projects/search' do
+      it 'should return search result in format json' do
```
Comment (igorkasyanchuk) [inline spec/controllers/account/projects_controller_spec.rb]: test is not complete

your DB is blank (are you sure it returns something?)

Comment (igorkasyanchuk) [reply spec/controllers/account/projects_controller_spec.rb]: at least not empty response

File: spec/models/project_spec.rb [hunk 11]
```diff
@@ -37,12 +37,18 @@
...
+    let!(:project1) { create(:project, name: 'Asana Demo', workspace_id: workspace.id) }
+    let!(:project2) { create(:project, name: 'Facebook', workspace_id: workspace.id) }
+    let!(:project3) { create(:project, name: 'Faker', workspace_id: workspace.id) }
 
     it 'should sort data by desc' do
       expect(workspace.projects.order_desc).to eq [project3, project2, project1]
     end
+
+    it 'should find all projects with entered text' do
+      expect(workspace.projects.search_projects('Fa').count).to eq 2
```
Comment (igorkasyanchuk) [inline spec/models/project_spec.rb:49]: good ... this test is better. I see it returns something

File: app/views/layouts/application.html.slim [hunk 12]
```diff
@@ -17,3 +17,5 @@ html
       = render 'shared/footer'
 
     = any_login_here if defined?(AnyLogin)
+    javascript:
+      document.addEventListener('DOMContentLoaded', global_search());
```
Comment (igorkasyanchuk) [inline app/views/layouts/application.html.slim]: and why it's here? specify it in JS itself

File: app/assets/javascripts/search.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
+function global_search() {
+  search.addEventListener('keyup', e => {
+
+    var search = document.getElementById('search');
+    var searchResults = document.getElementById('search-results');
+
+    var workspaces = document.getElementById('workspaces');
```
Comment (romaslmd) [inline app/assets/javascripts/search.js]: why plural?

File: config/routes.rb (hunk 9 above)
Comment (romaslmd) [reply config/routes.rb]: resource-like

File: spec/controllers/account/projects_controller_spec.rb (hunk 10 above)
Comment (romaslmd) [reply spec/controllers/account/projects_controller_spec.rb]: json requests should be validated
better to move them into a requests tests

Comment (igorkasyanchuk) [reply spec/controllers/account/projects_controller_spec.rb]: Roman what is "request test"? 

Maybe let's not overcomplicate things :) as suggested add validation to check that keys in response are the same as expected.

File: app/assets/javascripts/search.js (hunk 13 above)
Comment (BohdanBusko) [reply app/assets/javascripts/search.js]: because it includes not only one workspace

File: app/assets/javascripts/search.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
+document.addEventListener('DOMContentLoaded', () => {
+  globalSearch();
+});
+
+document.addEventListener('turbolinks:load', () => {
+  globalSearch();
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/search.js:2]: ми можемо лишити тільки turbolinks:load ??

що без DOMContentLoaded не працює? просто 2 рази викликається ця фунція і не знаю навіщо

File: spec/controllers/account/search_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  it 'should be success' do
+    get :index
+
+    expect(response).to be_successful
+  end
+
+  it 'should return json' do
+    get :index, params: { search: 'work' }
+
+    expect(JSON.parse(response.body)).to be_include('workspaces')
```
Comment (igorkasyanchuk) [inline spec/controllers/account/search_controller_spec.rb]: тест не гуд
перевір що ключі респонса ті які ти хочеш

і проблема ще що ти 3 рази викликаєш JSON.parse(response.body) ... навіщо копіпастати код. А якщо там 10 мб даних в респонсі то буде 3х повільніше працювати цей код

DRY code

File: app/models/user.rb [hunk 16]
```diff
@@ -87,6 +87,12 @@ def watching?(task)
     self.tasks.where(id: task.id).exists?
   end
 
+  def search_tasks(search)
+    self.projects.each do |project|
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: this code sucks

rework and use associations and READ ABOUT EACH / RETURN / MAP

it's strange to me this code

You NEED TO SEACH IN SCOPE OF ALL TASKS directly without loop by projects

File: app/models/user.rb [hunk 17]
```diff
@@ -87,6 +87,12 @@ def watching?(task)
...
+      return project.tasks.where("title ~* ?", "\\m#{search}").map
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: where is some limits in results? If I type "a" I will get 100000 results from DB. return only 10 per type/search

File: app/controllers/account/search_controller.rb [hunk 18]
```diff
@@ -0,0 +1,9 @@
+class Account::SearchController < Account::AccountController
+  def index
+    render json: {
+      workspaces: current_user.workspaces.search_workspaces(params[:search]),
```
Comment (igorkasyanchuk) [inline app/controllers/account/search_controller.rb:4]: what you return in JSON? the whole objects?
why?
return fields only which you need.
I don't think you need anything else except "id" and "title" correct?

**Review by igorkasyanchuk — APPROVED** [review]

**Pull request merged**